File: src/main/java/io/pageobject/generator/locator/protractor/IdLocatorSource.java
package io.pageobject.generator.locator.protractor;

import io.pageobject.generator.GeneratorContext;
import io.pageobject.generator.attribute.HtmlAttributes;
import io.pageobject.generator.locator.AbstractLocatorSource;
import org.jsoup.nodes.Element;

import static io.pageobject.generator.attribute.Attributes.extractAttributeValueAllowingIndexExpression;

public class IdLocatorSource extends AbstractLocatorSource {

    @Override
    public String extractLocatorValue(Element element, GeneratorContext context) {
        return extractAttributeValueAllowingIndexExpression(element, HtmlAttributes.ID);
    }

    public String locatorTemplate() {
        return "by.id('%s')";
    }

}

File: lib/locator/protractor/id-locator-source.js
"use strict";

const AbstractLocatorSource = require('../abstract-locator-source');
const _ = require('lodash');
const HtmlAttribute = require('../../attribute/html-attribute');
const extractAttributeValueAllowingIndexExpression = require('../../attribute/attributes').extractAttributeValueAllowingIndexExpression;

const template = _.template("by.id('<%= value %>')");

class IdLocatorSource extends AbstractLocatorSource {

  extractLocatorValue(context) {
    return extractAttributeValueAllowingIndexExpression(context.element, HtmlAttribute.ID);
  }

  locatorTemplate() {
    return template;
  }

}

module.exports = IdLocatorSource;